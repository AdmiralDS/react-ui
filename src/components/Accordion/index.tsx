import type { FC, HTMLAttributes, KeyboardEvent } from 'react';
import { useCallback, useRef } from 'react';

import { keyboardKey } from '../common/keyboardKey';
import { moveFocus, nextItem, previousItem } from './utils';
import { AccordionWrapper } from './style';

type Dimension = 'l' | 'm';

export * from './AccordionItem';

export interface AccordionProps extends HTMLAttributes<HTMLDivElement> {
  /** Размер компонента */
  dimension?: Dimension;
  /** Расположение иконки шеврона в заголовке. По умолчанию иконка выравнивается по правому краю. */
  iconPosition?: 'right' | 'left';
  /** Скрыть верхний разделитель аккордиона (верхнюю серую полоску).
   * Параметр hideTopDivider скрывает верхнюю серую полоску только у первого AccordionItem, заданного внутри Accordion
   */
  hideTopDivider?: boolean;
  /** Скрыть нижний разделитель аккордиона (нижнюю серую полоску)
   * Параметр hideBottomDivider скрывает нижнюю серую полоску только у последнего AccordionItem, заданного внутри Accordion
   */
  hideBottomDivider?: boolean;
}

export const Accordion: FC<AccordionProps> = ({
  children,
  dimension = 'l',
  iconPosition = 'right',
  hideTopDivider = false,
  hideBottomDivider = false,
  onKeyDown,
  ...props
}) => {
  const accordionRef = useRef<HTMLDivElement | null>(null);
  const handleKeyDown = useCallback((e: KeyboardEvent<HTMLDivElement>) => {
    let focusedOption: any = ((accordionRef.current && accordionRef.current.ownerDocument) || document).activeElement;
    // если фокус находится на AccordionItem Title, то можно переключаться между accordion items с помощью ArrowDown, ArrowUp, Home, End
    if (focusedOption.id.indexOf('accordion_title') > -1 && focusedOption.tagName.toLowerCase() === 'div') {
      do {
        focusedOption = focusedOption?.parentNode;
      } while (focusedOption.parentNode !== accordionRef.current);

      const code = keyboardKey.getCode(e);

      if (code === keyboardKey.ArrowDown) {
        moveFocus(accordionRef.current, focusedOption, nextItem);
        e.preventDefault();
      } else if (code === keyboardKey.ArrowUp) {
        moveFocus(accordionRef.current, focusedOption, previousItem);
        e.preventDefault();
      } else if (code === keyboardKey.Home) {
        e.preventDefault();
        moveFocus(accordionRef.current, null, nextItem);
      } else if (code === keyboardKey.End) {
        e.preventDefault();
        moveFocus(accordionRef.current, null, previousItem);
      }
    }
    onKeyDown?.(e);
  }, []);
  return (
    <AccordionWrapper
      ref={accordionRef}
      data-dimension={dimension}
      data-icon={iconPosition}
      $hideTopDivider={hideTopDivider}
      $hideBottomDivider={hideBottomDivider}
      $dimension={dimension}
      onKeyDown={handleKeyDown}
      {...props}
    >
      {children}
    </AccordionWrapper>
  );
};

Accordion.displayName = 'Accordion';
