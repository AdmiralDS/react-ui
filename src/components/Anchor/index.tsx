import type { HTMLAttributes, RefObject } from 'react';
import { createRef, forwardRef, useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { AnchorContainer } from './styled';
import type { AnchorDimension } from './AnchorItem';
import { AnchorItem } from './AnchorItem';
import { ActiveVerticalSelector } from './ActiveVerticalSelector';
import { getInternalCurrentAnchor } from './utils';
import { debounce } from '../common/utils/debounce';
import { refSetter } from '../common/utils/refSetter';

export { AnchorDimension };

export interface AnchorLinkItemProps {
  key: string;
  href: string;
  title: string;
  children?: Array<AnchorLinkItemProps>;
}

export interface AnchorProps extends HTMLAttributes<HTMLDivElement> {
  /** Размер компонента */
  dimension?: AnchorDimension;
  /** Сдвиг сверху при расчете положения прокрутки, в пикселях */
  offsetTop?: number;
  /** Границы зоны якоря, в пикселях */
  bounds?: number;
  /** Многострочное отображение текста, по умолчанию false */
  multilineView?: boolean;
  /** Элементы для отображения */
  items: Array<AnchorLinkItemProps>;
  /** Элемент для отслеживания скролла */
  getAnchorContainer?: () => HTMLElement | Window;
}

function getDefaultContainer() {
  return window;
}

interface NodesMapItem extends Omit<AnchorLinkItemProps, 'children'> {
  parent?: string;
  level: number;
}
type NodesMapItemWithRefProps = NodesMapItem & { ref: RefObject<HTMLAnchorElement> };

const itemHasChildren = (item: AnchorLinkItemProps) =>
  !!item.children && Array.isArray(item.children) && item.children.length > 0;

const treeToFlat = (tree: Array<AnchorLinkItemProps>, level = 0, parent?: string): Array<NodesMapItem> => {
  return tree.reduce((acc: Array<NodesMapItem>, item) => {
    const key = item.key.toString();
    const { href, title } = item;
    acc.push({ key, href, title, parent, level });
    if (item.children && itemHasChildren(item)) {
      const map = treeToFlat(item.children, level + 1, key);
      return [...acc, ...map];
    }

    return acc;
  }, []);
};

export const Anchor = forwardRef<HTMLDivElement, AnchorProps>(
  ({ dimension = 'm', offsetTop = 0, bounds = 5, multilineView = false, items, getAnchorContainer, ...props }, ref) => {
    const getCurrentContainer = getAnchorContainer ?? getDefaultContainer;
    const itemsMap = useMemo(() => treeToFlat(items), [items]);
    // add refs to links
    const itemsWithRef: Array<NodesMapItemWithRefProps> = useMemo(() => {
      return itemsMap.map((item) => ({ ...item, ref: createRef<HTMLAnchorElement>() }));
    }, [itemsMap]);

    const [activeLink, setActiveLink] = useState<string | null>(itemsMap[0]?.href || null);

    const handleScroll: EventListener = useCallback(() => {
      // TODO: add stop while animating
      // if (animating.current) {
      //   return;
      // }
      const links = itemsMap.map(({ href }) => href);
      const container = getCurrentContainer();
      const activeLink = getInternalCurrentAnchor(container, links, offsetTop, bounds);

      setActiveLink(activeLink);
    }, [getAnchorContainer, itemsMap]);

    //<editor-fold desc="Параметры для корректной отрисовки ActiveVerticalSelector">
    const anchorContainerRef = useRef<HTMLDivElement | null>(null);
    const [selectorTop, setSelectorTop] = useState(0);
    const [selectorHeight, setSelectorHeight] = useState(0);
    const selectorRef = useRef<HTMLDivElement | null>(null);

    const styleSelector = (top: number, height: number) => {
      setSelectorTop(top);
      setSelectorHeight(height);
    };

    useEffect(() => {
      function setSelector() {
        const activeLinkRef = itemsWithRef.filter((item) => item.href === activeLink)?.[0]?.ref.current;
        const top = parseFloat(selectorRef.current?.style.top || '0');
        const selectorHeight = parseFloat(selectorRef.current?.style.height || '0');

        if (activeLinkRef && anchorContainerRef.current) {
          // используем метод getBoundingClientRect, так как он дает точность до сотых пикселя
          const activeLinkHeight = activeLinkRef.getBoundingClientRect().height;
          const activeLinkTop =
            activeLinkRef.getBoundingClientRect().top -
            anchorContainerRef.current.getBoundingClientRect().top +
            anchorContainerRef.current.scrollTop;

          if (activeLinkTop !== top || activeLinkHeight !== selectorHeight) {
            styleSelector(activeLinkTop, activeLinkHeight);
          }
        }
        if (!activeLinkRef) {
          styleSelector(0, 0);
        }
      }

      if (anchorContainerRef.current?.firstElementChild) {
        const resizeObserver = new ResizeObserver(debounce(setSelector, 100));
        resizeObserver.observe(anchorContainerRef.current?.firstElementChild);
        return () => {
          resizeObserver.disconnect();
        };
      }
    }, [itemsWithRef, activeLink, dimension, multilineView, offsetTop, bounds]);
    //</editor-fold>

    useEffect(() => {
      const container = getCurrentContainer();
      container.addEventListener('scroll', handleScroll);
      return () => {
        container.removeEventListener('scroll', handleScroll);
      };
    }, [getAnchorContainer, handleScroll]);

    const renderItems = () => {
      return itemsWithRef.map((item) => (
        <AnchorItem
          ref={item.ref}
          key={item.key}
          dimension={dimension}
          multilineView={multilineView}
          href={item.href}
          active={item.href === activeLink}
          level={item.level}
        >
          {item.title}
        </AnchorItem>
      ));
    };

    return (
      <AnchorContainer {...props} ref={refSetter(ref, anchorContainerRef)}>
        {renderItems()}
        <ActiveVerticalSelector $top={selectorTop} $height={selectorHeight} $transition={true} />
      </AnchorContainer>
    );
  },
);
Anchor.displayName = 'Anchor';
