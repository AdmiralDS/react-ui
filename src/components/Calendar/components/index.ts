export * from './Years';
export * from './Button';
export * from './Day';
export * from './DayNames';
export * from './Month';
export * from './Months';
export * from './Panel';
export * from './Week';
