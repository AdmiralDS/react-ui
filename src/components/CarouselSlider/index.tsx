import type { HTMLAttributes } from 'react';
import styled, { css } from 'styled-components';

export type CarouselSliderAppearance = 'default' | 'primary';

export interface CarouselSliderItemProps extends HTMLAttributes<HTMLButtonElement> {
  /** Внешний вид компонента */
  appearance?: CarouselSliderAppearance;
  /** Секция является активной */
  isCurrent?: boolean;
}
const primaryAppearanceMixin = css`
  background-color: var(--admiral-color-Neutral_Neutral30, ${(p) => p.theme.color['Neutral/Neutral 30']});
  border: none;
`;
const primaryAppearanceSelectedMixin = css`
  background-color: var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']});
  border: none;
`;
const defaultAppearanceMixin = css`
  background-color: var(--admiral-color-Special_StaticWhite, ${(p) => p.theme.color['Special/Static White']});
  border: 0.5px solid
    var(--admiral-color-Special_DarkStaticNeutral00, ${(p) => p.theme.color['Special/Dark Static Neutral 80']});
`;
const ItemContent = styled.div<{ $appearance?: CarouselSliderAppearance; $isCurrent?: boolean }>`
  box-sizing: border-box;
  width: ${(p) => (p.$isCurrent ? 28 : 12)}px;
  height: 4px;
  border-radius: 2px;
  ${(p) =>
    p.$appearance === 'default'
      ? defaultAppearanceMixin
      : p.$isCurrent
        ? primaryAppearanceSelectedMixin
        : primaryAppearanceMixin};
`;
const Item = styled.button<{ $isCurrent?: boolean }>`
  all: unset;
  display: block;
  appearance: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;

  box-sizing: border-box;
  width: ${(p) => (p.$isCurrent ? 34 : 18)}px;
  height: 16px;
  padding: 7px 3px 6px 3px;

  &:focus-visible > ${ItemContent} {
    outline-offset: 2px;
    outline: var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']}) solid 2px;
  }
`;

export const CarouselSliderItem = ({
  appearance = 'default',
  isCurrent = false,
  ...props
}: CarouselSliderItemProps) => {
  return (
    <Item {...props} $isCurrent={isCurrent}>
      <ItemContent $appearance={appearance} $isCurrent={isCurrent} />
    </Item>
  );
};
CarouselSliderItem.displayName = 'CarouselSliderItem';

const Wrapper = styled.div`
  position: relative;
  display: flex;
  overflow-x: scroll;
  width: fit-content;

  &::-webkit-scrollbar {
    width: 0 !important;
    height: 0 !important;
  }

  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  scrollbar-width: none;

  & ${Item}, ${ItemContent} {
    transition: width 0.3s ease-in-out;
  }
`;

export interface CarouselSliderProps extends HTMLAttributes<HTMLDivElement> {}

export const CarouselSlider = ({ children, ...props }: CarouselSliderProps) => {
  return <Wrapper {...props}>{children}</Wrapper>;
};
CarouselSlider.displayName = 'CarouselSlider';
