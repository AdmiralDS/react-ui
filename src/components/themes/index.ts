export * from './light';
export * from './dark';
export type { Theme } from './common';
