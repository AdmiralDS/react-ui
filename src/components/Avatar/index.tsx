import { forwardRef } from 'react';

import type { AvatarBaseInternalProps, AvatarBaseProps } from '#src/components/AvatarBase';
import { AvatarBase } from '#src/components/AvatarBase';

export interface AvatarProps extends Omit<AvatarBaseProps, 'withActivityRing' | 'showActivityRing'> {}

export const Avatar = forwardRef<HTMLButtonElement, AvatarProps & AvatarBaseInternalProps>(
  ({ dimension, ...props }: AvatarProps & AvatarBaseInternalProps, ref) => {
    return (
      <AvatarBase
        ref={ref}
        {...props}
        dimension={props.status && dimension === 'xxs' ? 'xs' : dimension}
        withActivityRing={false}
        showActivityRing={false}
      />
    );
  },
);

Avatar.displayName = 'Avatar';
