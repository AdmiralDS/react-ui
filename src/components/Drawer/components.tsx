import type { FC, HTMLAttributes } from 'react';
import { createContext, useContext, useLayoutEffect, useRef } from 'react';
import styled from 'styled-components';

import { typography } from '#src/components/Typography';
import { debounce } from '#src/components/common/utils/debounce';
import { resizePaddings } from '../common/dom/resizePaddings';

const Title = styled.h5<{ $mobile: boolean; $displayCloseIcon: boolean }>`
  ${({ $mobile }) => ($mobile ? typography['Header/H6'] : typography['Header/H5'])};
  color: var(--admiral-color-Neutral_Neutral90, ${(p) => p.theme.color['Neutral/Neutral 90']});
  margin: 0;
  padding: ${({ $mobile, $displayCloseIcon }) => {
    if ($mobile) {
      return $displayCloseIcon ? '0 48px 8px 16px' : '0 16px 8px';
    }
    return $displayCloseIcon ? '0 56px 8px 24px' : '0 24px 8px';
  }};
`;

const Content = styled.div<{ $mobile: boolean }>`
  overflow-y: auto;
  outline: none;
  scrollbar-gutter: stable;
  padding-block: 8px;
  padding-inline: ${({ $mobile }) => `${$mobile ? 16 : 24}px`};
  height: 100%;
`;

const ButtonPanel = styled.div<{ $mobile: boolean }>`
  display: flex;
  flex-direction: ${({ $mobile }) => ($mobile ? 'column-reverse' : 'row-reverse')};
  padding: ${({ $mobile }) => ($mobile ? '16px 16px 0' : '16px 24px 0')};

  & > button {
    margin: ${({ $mobile }) => ($mobile ? '0 0 16px 0' : '0 16px 0 0')};
    ${({ $mobile }) => $mobile && 'width: 100%;'}
  }

  & > button:first-child {
    margin: 0;
  }
`;

export const DrawerContext = createContext({ mobile: false, displayCloseIcon: true } as {
  mobile: boolean;
  displayCloseIcon: boolean;
});

export const DrawerTitle: FC<HTMLAttributes<HTMLHeadingElement>> = ({ children, ...props }) => {
  const { mobile, displayCloseIcon } = useContext(DrawerContext);
  const asProp = mobile ? 'h6' : 'h5';
  return (
    <Title $mobile={mobile} $displayCloseIcon={displayCloseIcon} as={asProp} {...props}>
      {children}
    </Title>
  );
};

export const DrawerContent: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const contentRef = useRef<HTMLDivElement | null>(null);
  const mobile = useContext(DrawerContext).mobile;

  useLayoutEffect(() => {
    const node = contentRef.current;
    if (node) {
      resizePaddings(node);
      const resizeObserver = new ResizeObserver(debounce(() => resizePaddings(node), 250));
      resizeObserver.observe(node);
      return () => {
        resizeObserver.unobserve(node);
      };
    }
  }, []);

  return (
    <Content tabIndex={-1} ref={contentRef} $mobile={mobile} {...props}>
      {children}
    </Content>
  );
};

export const DrawerButtonPanel: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const mobile = useContext(DrawerContext).mobile;
  return (
    <ButtonPanel $mobile={mobile} {...props}>
      {children}
    </ButtonPanel>
  );
};
