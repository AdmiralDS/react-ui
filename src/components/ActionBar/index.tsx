import { Fragment, useLayoutEffect, useMemo, useRef, useState } from 'react';
import styled from 'styled-components';

import type { ActionBarDimension, ActionBarItemProps, ActionBarProps } from '#src/components/ActionBar/types';
import {
  ACTION_BAR_ITEM_HEIGHT_L,
  ACTION_BAR_ITEM_HEIGHT_M,
  ACTION_BAR_ITEM_HEIGHT_S,
  ACTION_BAR_ITEM_HEIGHT_XL,
  DIVIDER_SIZE_M,
  DIVIDER_SIZE_S,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_L,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_M,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_S,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_XL,
} from '#src/components/ActionBar/constants';
import { OverflowMenu } from '#src/components/OverflowMenu';
import { ActionBarDivider } from '#src/components/ActionBar/ActionBarItems';
import type { MenuModelItemProps } from '#src/components/Menu/MenuItem';
import { mediumGroupBorderRadius } from '#src/components/themes';

export * from './types';
export * from './ActionBarItems';

const getActionBarHeight = (dimension: ActionBarDimension) => {
  switch (dimension) {
    case 's':
      return ACTION_BAR_ITEM_HEIGHT_S;
    case 'm':
      return ACTION_BAR_ITEM_HEIGHT_M;
    case 'l':
      return ACTION_BAR_ITEM_HEIGHT_L;
    case 'xl':
    default:
      return ACTION_BAR_ITEM_HEIGHT_XL;
  }
};

const Wrapper = styled.div<{ $dimension: ActionBarDimension }>`
  box-sizing: border-box;
  height: ${({ $dimension }) => getActionBarHeight($dimension)}px;

  display: flex;
  align-items: center;
`;
const ActionBarOverflowMenuItem = styled.div<{ $dimension: ActionBarDimension }>`
  flex-shrink: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: ${({ $dimension }) => getActionBarHeight($dimension)}px;
  height: ${({ $dimension }) => getActionBarHeight($dimension)}px;
`;
const getFocusBorderOffset = ($dimension: ActionBarDimension) => {
  switch ($dimension) {
    case 's':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_S;
    case 'm':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_M;
    case 'l':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_L;
    case 'xl':
    default:
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_XL;
  }
};

function getFocusBorderRadius(componentBorderRadius: string) {
  switch (componentBorderRadius) {
    case '0':
      return '0';
    case '2px':
      return '6px';
    case '6px':
      return '10px';
    case '8px':
      return '12px';
    case '10px':
      return '14px';
    case '4px':
    default:
      return '8px';
  }
}

const StyledOverflowMenu = styled(OverflowMenu)<{ $dimension: ActionBarDimension }>`
  &:focus-visible:not(:disabled) {
    outline: none;

    &:before {
      border: 2px solid var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']});
      border-radius: ${(p) => getFocusBorderRadius(mediumGroupBorderRadius(p.theme.shape))};
      content: '';
      display: block;
      position: absolute;
      width: ${({ $dimension }) => getActionBarHeight($dimension) + 2 * 2}px;
      height: ${({ $dimension }) => getActionBarHeight($dimension) + 2 * 2}px;
      top: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      left: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      right: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      bottom: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      pointer-events: none;
    }
  }
`;

export const ActionBar = ({
  dimension = 'xl',
  renderActionBarItem,
  renderDropMenuItem,
  items,
  itemIsDisabled,
  ...props
}: ActionBarProps) => {
  const renderItem = ({ itemId, withDivider }: ActionBarItemProps) => {
    return (
      <Fragment key={itemId}>
        {renderActionBarItem(itemId)}
        {withDivider && <ActionBarDivider key={`divider${itemId}`} dimension={dimension} />}
      </Fragment>
    );
  };

  const [containerWidth, setContainerWidth] = useState(0);
  const containerRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    if (containerRef.current) {
      const resizeObserver = new ResizeObserver((entries) => {
        entries.forEach((entry) => setContainerWidth(entry.contentRect.width || 0));
      });
      resizeObserver.observe(containerRef.current);
      return () => {
        resizeObserver.disconnect();
      };
    }
  }, []);
  const [visibleItems, setVisibleItems] = useState<ActionBarItemProps[]>([]);
  const [hiddenItems, setHiddenItems] = useState<ActionBarItemProps[]>([]);

  useLayoutEffect(() => {
    if (containerRef.current) {
      const buttonWidth = getActionBarHeight(dimension);
      const dividerWidth = dimension === 's' ? DIVIDER_SIZE_S : DIVIDER_SIZE_M;

      const totalItemsWidth = items.reduce((acc, item) => {
        return acc + buttonWidth + (item.withDivider ? dividerWidth : 0);
      }, 0);
      const overflow = totalItemsWidth > containerWidth;

      let availableWidth = containerWidth - (overflow ? buttonWidth : 0);

      const newVisibleItems: ActionBarItemProps[] = [];
      const newHiddenItems: ActionBarItemProps[] = [];
      const addToVisible = (item: ActionBarItemProps) => newVisibleItems.push(item);
      const addToHidden = (item: ActionBarItemProps) => newHiddenItems.push(item);

      items.forEach((item) => {
        const itemWidth = buttonWidth + (item.withDivider ? dividerWidth : 0);
        if (availableWidth >= itemWidth) {
          addToVisible(item);
          availableWidth -= itemWidth;
        } else {
          addToHidden(item);
          availableWidth = 0;
        }
      });

      setVisibleItems(newVisibleItems);
      setHiddenItems(newHiddenItems);
    }
  }, [items, containerWidth, dimension]);
  const overflowMenuItems: MenuModelItemProps[] = useMemo(() => {
    if (hiddenItems.length === 0) return [];
    return hiddenItems.map(({ itemId }) => {
      return {
        id: itemId,
        render: renderDropMenuItem(itemId),
        disabled: itemIsDisabled(itemId),
      };
    });
  }, [hiddenItems, dimension]);

  return (
    <Wrapper {...props} ref={containerRef} $dimension={dimension}>
      {visibleItems.map((item) => renderItem(item))}
      {hiddenItems.length > 0 && (
        <ActionBarOverflowMenuItem $dimension={dimension}>
          <StyledOverflowMenu
            $dimension={dimension}
            items={overflowMenuItems}
            dimension={dimension === 'xl' ? 'l' : dimension}
          />
        </ActionBarOverflowMenuItem>
      )}
    </Wrapper>
  );
};
