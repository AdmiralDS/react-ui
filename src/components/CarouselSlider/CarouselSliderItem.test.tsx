import { fireEvent, render, screen } from '@testing-library/react';
import { LIGHT_THEME } from '#src/components/themes';
import { ThemeProvider } from 'styled-components';
import { CarouselSliderItem } from '.';

describe('CarouselSliderItem', () => {
  const carouselSliderItemRequiredProps = {
    onClick: jest.fn(),

    'aria-label': 'testItem',
    'data-testid': 'testItem',
  };

  it('should render item with default appearance', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <CarouselSliderItem {...carouselSliderItemRequiredProps} key={0} />
      </ThemeProvider>,
    );

    const itemElement = screen.getByTestId('testItem');
    expect(itemElement).toMatchSnapshot();
  });

  it('should render item with primary appearance', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <CarouselSliderItem {...carouselSliderItemRequiredProps} key={0} appearance="primary" />
      </ThemeProvider>,
    );

    const itemElement = screen.getByTestId('testItem');
    expect(itemElement).toMatchSnapshot();
  });

  it('should render current item with default appearance', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <CarouselSliderItem {...carouselSliderItemRequiredProps} key={0} isCurrent />
      </ThemeProvider>,
    );

    const itemElement = screen.getByTestId('testItem');
    expect(itemElement).toMatchSnapshot();
  });

  it('should render current item with primary appearance', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <CarouselSliderItem {...carouselSliderItemRequiredProps} key={0} appearance="primary" isCurrent />
      </ThemeProvider>,
    );

    const itemElement = screen.getByTestId('testItem');
    expect(itemElement).toMatchSnapshot();
  });

  it('should call onClick on item', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <CarouselSliderItem {...carouselSliderItemRequiredProps} key={0} />
      </ThemeProvider>,
    );

    const itemElement = screen.getByTestId('testItem');
    fireEvent.click(itemElement);
    const { onClick } = carouselSliderItemRequiredProps;
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
