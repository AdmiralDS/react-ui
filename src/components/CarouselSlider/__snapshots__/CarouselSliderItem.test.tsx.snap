// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSliderItem should render current item with default appearance 1`] = `
.c2 {
  box-sizing: border-box;
  width: 28px;
  height: 4px;
  border-radius: 2px;
  background-color: var(--admiral-color-Special_StaticWhite, #FFFFFF);
  border: 0.5px solid var(--admiral-color-Special_DarkStaticNeutral00, #C9CACD);
}

.c0 {
  all: unset;
  display: block;
  appearance: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  box-sizing: border-box;
  width: 34px;
  height: 16px;
  padding: 7px 3px 6px 3px;
}

.c0:focus-visible>.c1 {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

<button
  aria-label="testItem"
  class="c0"
  data-testid="testItem"
>
  <div
    class="c1 c2"
  />
</button>
`;

exports[`CarouselSliderItem should render current item with primary appearance 1`] = `
.c2 {
  box-sizing: border-box;
  width: 28px;
  height: 4px;
  border-radius: 2px;
  background-color: var(--admiral-color-Primary_Primary60Main, #0062FF);
  border: none;
}

.c0 {
  all: unset;
  display: block;
  appearance: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  box-sizing: border-box;
  width: 34px;
  height: 16px;
  padding: 7px 3px 6px 3px;
}

.c0:focus-visible>.c1 {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

<button
  aria-label="testItem"
  class="c0"
  data-testid="testItem"
>
  <div
    class="c1 c2"
  />
</button>
`;

exports[`CarouselSliderItem should render item with default appearance 1`] = `
.c2 {
  box-sizing: border-box;
  width: 12px;
  height: 4px;
  border-radius: 2px;
  background-color: var(--admiral-color-Special_StaticWhite, #FFFFFF);
  border: 0.5px solid var(--admiral-color-Special_DarkStaticNeutral00, #C9CACD);
}

.c0 {
  all: unset;
  display: block;
  appearance: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  box-sizing: border-box;
  width: 18px;
  height: 16px;
  padding: 7px 3px 6px 3px;
}

.c0:focus-visible>.c1 {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

<button
  aria-label="testItem"
  class="c0"
  data-testid="testItem"
>
  <div
    class="c1 c2"
  />
</button>
`;

exports[`CarouselSliderItem should render item with primary appearance 1`] = `
.c2 {
  box-sizing: border-box;
  width: 12px;
  height: 4px;
  border-radius: 2px;
  background-color: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
  border: none;
}

.c0 {
  all: unset;
  display: block;
  appearance: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  box-sizing: border-box;
  width: 18px;
  height: 16px;
  padding: 7px 3px 6px 3px;
}

.c0:focus-visible>.c1 {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

<button
  aria-label="testItem"
  class="c0"
  data-testid="testItem"
>
  <div
    class="c1 c2"
  />
</button>
`;
